### PR TITLE
Allow ASOF JOIN over nullable right column

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):

- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
* Allow ASOF JOIN over nullable right column

This PR is also fixes `01428_nullable_asof_join` for analyzer, but change is for both old and new one.

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
This is an automated comment for commit 48ac6e3cea8ff50e6c8c375bc5cacb95a1132edb with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/49826/48ac6e3cea8ff50e6c8c375bc5cacb95a1132edb/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Interpreters/HashJoin.cpp [hunk 1]
```diff
@@ -704,15 +704,46 @@ Block HashJoin::prepareRightBlock(const Block & block) const
...
-        throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Too many rows in right table block for HashJoin: {}", source_block.rows());
+    if (unlikely(source_block_.rows() > std::numeric_limits<RowRef::SizeT>::max()))
+        throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Too many rows in right table block for HashJoin: {}", source_block_.rows());
+
+    Block source_block = source_block_;
+    if (strictness == JoinStrictness::Asof)
+    {
+        chassert(kind == JoinKind::Left || kind == JoinKind::Inner);
+
+        // Filter out rows with NULLs in asof key
```
Comment (novikd) [inline src/Interpreters/HashJoin.cpp:722]: Better to add a comment on why we should do it.

**Review by novikd — APPROVED** [review]

File: tests/broken_tests.txt [hunk 2]
```diff
@@ -37,8 +37,6 @@
 01268_shard_avgweighted
 01270_optimize_skip_unused_shards_low_cardinality
 01319_optimize_skip_unused_shards_nesting
```
Comment (vdimir) [inline tests/broken_tests.txt:39]:
```suggestion
01319_optimize_skip_unused_shards_nesting
01353_low_cardinality_join_types
```

**Comment by vdimir** [comment]
> Mergeable Check — failed: Unit tests (asan); succeeded: ...
> Unit tests (asan) — PASSED record not found. [Details](https://s3.amazonaws.com/clickhouse-test-reports/49826/55746122cd4ea36bdfad8a10b5e2d66069b788b9/unit_tests__asan_.html)

Check is broken

https://play.clickhouse.com/play?user=play#U0VMRUNUIHB1bGxfcmVxdWVzdF9udW1iZXIsIGNoZWNrX3N0YXJ0X3RpbWUsIGNoZWNrX25hbWUsIGNoZWNrX3N0YXR1cywgcmVwb3J0X3VybApGUk9NIGNoZWNrcwpXSEVSRSAxCiAgICAtLUFORCBwdWxsX3JlcXVlc3RfbnVtYmVyID0gMAogICAgQU5EIGNoZWNrX3N0YXR1cyAhPSAnc3VjY2VzcycKICAgIEFORCBjaGVja19uYW1lIGxpa2UgJyVVbml0IHRlc3RzJScKICAgIEFORCBjaGVja19zdGFydF90aW1lID49IG5vdygpIC0gSU5URVJWQUwgMSBEQVkKT1JERVIgQlkgY2hlY2tfc3RhcnRfdGltZSBkZXNjLCBjaGVja19uYW1lLCB0ZXN0X25hbWUKTElNSVQgMSBCWSBjaGVja19zdGFydF90aW1l

**Pull request merged**